csrc/kernel_cache.cpp

```diff
@@ -979,8 +982,34 @@ FusionKernelRuntime::FusionKernelRuntime(
   // Initialize the evaluator simplifer
   precomputed_values_ = std::make_unique<PrecomputedValues>(fusion.get());
 
-  segmented_fusion_ =
-      SegmentCandidateFinder::segment(std::move(fusion), args, runtime_info);
+  if (serde_buffer == nullptr) {
+    // Default compilation path applies segmentation before scheduling and
+    // compiling the fusion.
+    segmented_fusion_ =
+        SegmentCandidateFinder::segment(std::move(fusion), args, runtime_info);
+  } else {
+    // Serialization path that generates segmented fusion from flatbuffers.
+    // Convert Welford to two-pass if option is enabled and the original
+    // heuristic is persistent
+    const flatbuffers::Vector<flatbuffers::Offset<serde::SegmentedGroup>>*
+        segmented_groups = serde_buffer->segmented_fusion()->groups();
+    bool has_persistent_heuristic = std::any_of(
+        segmented_groups->begin(),
+        segmented_groups->end(),
+        [](const serde::SegmentedGroup* sg) {
+          auto heuristic = static_cast<ScheduleHeuristic>(sg->heuristic());
+          return heuristic == ScheduleHeuristic::InnerPersistent ||
+              heuristic == ScheduleHeuristic::OuterPersistent ||
+              heuristic == ScheduleHeuristic::InnerOuterPersistent;
+        });
+
+    bool has_welford_ops = ir_utils::hasOpsOfType<WelfordOp>(fusion.get());
+    if (has_welford_ops && has_persistent_heuristic) {
+      SegmentCandidateFinder::translateWelfordInFusion(fusion.get(), args);
```

Comment:
We modify Fusion here because we don't serialize Fusion? Is there value (in the future) to serialize/deserialize fusion IR so we can cache fusion modifications as well?

Reply:
Serializing the fusion IR is planned but this welford transformation is the top priority.